File: libNative/src/main/java/com/munch/lib/libnative/helper/ViewHelper.java
package com.munch.lib.libnative.helper;

import android.support.annotation.NonNull;
import android.widget.TextView;

import com.munch.lib.libnative.exception.MethodException;

import java.util.ArrayList;
import java.util.List;

/**
 * 一些各类View中常用的方法
 * <p>
 * Created by Munch on 2018/12/26 14:25.
 */
public class ViewHelper {

    private ViewHelper() {
    }

    /**
     * 一般用于一个页面需要多个输入的值的判断和获取，注意，返回值是根据传入的顺序返回的
     * <p>
     * 对传入的TextView或其子类的getText()值进行判断，
     * 如不为空，已照TextView传入的顺序放入{@link OnCheckTextViewListener#onNoEmpty(List)}的参数中并在检查完后回调
     * 如有值为空，则回调{@link OnCheckTextViewListener#onCheckEmpty(TextView)}并根据其返回值判断是否跳出
     *
     * @param listener 处理回调
     * @param views    被判断和获取的Textview及其子类
     */
    public static void checkTextViewEmpty(@NonNull OnCheckTextViewListener listener, @NonNull TextView... views) {
        String val;
        List<String> vals = new ArrayList<>(views.length);
        for (TextView view : views) {
            val = view.getText().toString().trim();
            if (val.isEmpty() && listener.onCheckEmpty(view)) {
                break;
            }
            vals.add(val);
        }
        listener.onNoEmpty(vals);
    }

    /**
     * 对传入的TextView数值根据String数组进行赋值，两个数组需要对等
     * <p>
     * 若vals中有空值，则调用{@link ObjectHelper#requireNonNull(String)}使其不为空
     *
     * @param views 需要被设置值得TextView或其子类的数组
     * @param vals  值的数组
     */
    public static void setTextViewNonVal(TextView[] views, String[] vals) {
        int len = views.length;
        if (len != vals.length) {
            throw MethodException.defaultException("View与数据不对等");
        }
        for (int i = 0; i < len; i++) {
            views[i].setText(ObjectHelper.requireNonNull(vals[i]));
        }
    }

    /**
     * 对传入的TextView数值根据String数组进行赋值，两个数组需要对等
     * <p>
     * 未进行空值判断
     *
     * @param views 需要被设置值得TextView或其子类的数组
     * @param vals  值的数组
     */
    public static void setTextViewVal(TextView[] views, String[] vals) {
        int len = views.length;
        if (len != vals.length) {
            throw MethodException.defaultException("View与数据不对等");
        }
        for (int i = 0; i < len; i++) {
            views[i].setText(vals[i]);
        }
    }


    public interface OnCheckTextViewListener {

        /**
         * @param view 数据为空的Textview或其子类
         * @return 是否跳出，不再继续检查剩余的View
         */
        boolean onCheckEmpty(TextView view);

        /**
         * @param vals 按顺序获取的Textview或其子类的值的集合
         */
        void onNoEmpty(List<String> vals);
    }
}

File: libnative/src/main/java/com/munch/lib/libnative/helper/ViewHelper.kt
package com.munch.lib.libnative.helper

import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import androidx.core.view.children
import com.munch.lib.libnative.excetion.MethodException

/**
 * Created by Munch on 2019/7/15 13:50
 */
object ViewHelper {

    @JvmStatic
    fun tag4Once(view: View, run: (View) -> Unit) {
        val tag = view.tag as? Boolean?
        if (tag == null || !tag) {
            view.tag = true
            run(view)
        }
    }

    @JvmStatic
    fun addViewPadding(view: View, left: Int = 0, top: Int = 0, right: Int = 0, bottom: Int = 0) {
        view.setPadding(
            view.paddingLeft + left, view.paddingTop + top,
            view.paddingRight + right, view.paddingBottom + bottom
        )
    }

    @JvmStatic
    fun setViewPadding(view: View, left: Int = 0, top: Int = 0, right: Int = 0, bottom: Int = 0) {
        view.setPadding(left, top, right, bottom)
    }

    @JvmStatic
    fun setViewMargin(view: View, left: Int = 0, top: Int = 0, right: Int = 0, bottom: Int = 0) {
        val params = getParams(view)
        val marginParams: ViewGroup.MarginLayoutParams
        marginParams = if (params !is ViewGroup.MarginLayoutParams) {
            ViewGroup.MarginLayoutParams(params.width, params.height)
        } else {
            params
        }
        marginParams.setMargins(left, top, right, bottom)
        view.layoutParams = marginParams
    }

    @JvmStatic
    fun addViewMargin(view: View, left: Int = 0, top: Int = 0, right: Int = 0, bottom: Int = 0) {
        val params = getParams(view)
        val marginParams: ViewGroup.MarginLayoutParams
        marginParams = if (params !is ViewGroup.MarginLayoutParams) {
            ViewGroup.MarginLayoutParams(params.width, params.height)
        } else {
            params
        }
        marginParams.setMargins(
            marginParams.leftMargin + left, marginParams.topMargin + top,
            marginParams.rightMargin + right, marginParams.bottomMargin + bottom
        )
        view.layoutParams = marginParams
    }

    @JvmStatic
    fun setNonInput(editText: EditText) {
        editText.keyListener = null
        editText.isFocusable = false
    }

    @JvmStatic
    fun getParams(view: View): ViewGroup.LayoutParams {
        return view.layoutParams ?: ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
    }

    @JvmStatic
    fun clickItem(vg: ViewGroup, listener: View.OnClickListener, vararg clazz: Class<in View>) {
        vg.children.forEachIndexed { index, view ->
            if (clazz.isNotEmpty()) {
                clazz.forEach {
                    if (it.isAssignableFrom(view::class.java)) {
                        view.tag = index
                        view.setOnClickListener(listener)
                    }
                }
            } else {
                view.tag = index
                view.setOnClickListener(listener)
            }
        }

    }

    @JvmStatic
    fun setTextViewNonVal(views: Array<TextView>, strs: Array<String?>) {
        if (views.size != strs.size) {
            throw MethodException.wrongParameter()
        }
        views.forEachIndexed { index, textView ->
            textView.text = strs[index] ?: ""
        }
    }

    @JvmStatic
    fun checkTextEmpry(listener: OnCheckTextViewListener, vararg views: TextView) {
        if (views.isEmpty()) {
            return
        }
        val array = ArrayList<String>(views.size)
        views.forEach {
            val str = it.text.toString()
            if (str.isEmpty() && !listener.onCheckEmpty(it)) {
                return
            }
            array.add(str)
        }
        listener.onResult(array)
    }

    interface OnCheckTextViewListener {

        /**
         * 当检查到view的text为空时回调
         *
         * @param view 数据为空的Textview或其子类
         * @return 是否继续检查剩余的View，true则继续
         */
        fun onCheckEmpty(view: TextView): Boolean

        /**
         * 当检查完所有view都不为空或者有为空但返回true时回调
         * 注意vals的长度
         *
         * @param vals 按顺序获取的Textview或其子类的值的集合
         */
        fun onResult(vals: List<String>)
    }

}